### PR TITLE
dzsave: remove libgsf in favor of libarchive

### PR DESCRIPTION
**Pull request by kleisauke**
See: #3354.

Marked this PR as draft due to these TODO items:
- [x] ~Guard calls to libzip with a mutex, just like we did previously with libgsf?~ 2c5cfdb56a9800414aa49270a9242070d251d76d
- [x] ~Guard the whole `dzsave` operation with `#ifdef HAVE_ZIP`?~ 92d84163101c6c6401c130fa44f21edafdd13d04
- [x] ~Fix intermittent test failure on CI (unable to reproduce locally, perhaps a `mtime` issue?).~ 2c5cfdb56a9800414aa49270a9242070d251d76d
- [x] ~Check with Valgrind, LSan, etc.~ f8eaf631335f30151de91dbddd9e525667ba2934
- [x] ~Test ZIP64 support.~
  ```console
  $ vips dzsave CMU-2.svs[attach-associated] x.szi --overlap 0 --suffix .png[compression=0]
  $ du -h x.szi 
  12G	x.szi
  $ UNZIP_DISABLE_ZIPBOMB_DETECTION=TRUE zip -T x.szi
  test of x.szi OK
  ```
- [x] ~Test unseekable targets (e.g. pipes). https://github.com/nih-at/libzip/issues/378~ 0a2b20fa0caed226d8cc4977830d66507edecdbd
- [x] ~Benchmark?~ https://github.com/libvips/libvips/pull/3476#issuecomment-1581052370
- [x] Avoid empty directory when associated images are missing? e76251aedf13d6ea470ccef22a0223ce9563fd3c

### CODE REVIEW

File: libvips/foreign/dzsave.c [hunk 1]
```diff
@@ -866,61 +529,37 @@ write_image( VipsForeignSaveDz *dz,
...
-#ifndef HAVE_GSF_ZIP64
-	if( iszip( dz->container ) ) { 
-		/* Leave 3 entry headroom for blank.png and metadata files.
-		 */
-		if( dz->tree->file_count + 3 >= (unsigned int) USHRT_MAX ) {
-			g_mutex_unlock( vips__global_lock );
-
-			vips_error( class->nickname,
-				"%s", _( "too many files in zip" ) );
+		if( vips_image_write_to_buffer( t, dz->suffix, &buf, &len,
```
Comment (kleisauke) [inline libvips/foreign/dzsave.c:538]: This needs to be revisited, `blank.png` should always be written as PNG when `dz->suffix` is `".jpeg"`. The same applies to associated images, they should probably always be written as JPEG (i.e. we cannot assume that all image types are the same).

**Comment by jcupitt** [comment]
Very nice! 

> Guard calls to libzip with a mutex, just like we did previously with libgsf?

Yes, I think this is necessary. I tried:

```
$ vips dzsave k2.jpg x.szi
realloc(): invalid old size
Aborted (core dumped)
```

Which looks like a locking issue. With https://github.com/libvips/libvips/commit/2c5cfdb56a9800414aa49270a9242070d251d76d I see:

```
$ vips dzsave k2.jpg x.szi
$ file x.szi
x.szi: Zip archive data, at least v2.0 to extract, compression method=store
```

And it unzips correctly.

**Comment by jcupitt** [comment]
... which means zipping is single-threaded, but we will be using STORE 99% of the time, so I doubt if it matters.

**Comment by jcupitt** [comment]
I also tried:

```
$ valgrind --leak-check=yes vips dzsave k2.jpg x.szi
==790619== Memcheck, a memory error detector
==790619== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==790619== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==790619== Command: vips dzsave k2.jpg x.szi
==790619== 
memory: high-water mark 18.47 MB
==790619== 
==790619== HEAP SUMMARY:
==790619==     in use at exit: 2,298,424 bytes in 22,065 blocks
==790619==   total heap usage: 203,292 allocs, 181,227 frees, 65,802,463 bytes allocated
==790619== 
==790619== LEAK SUMMARY:
==790619==    definitely lost: 0 bytes in 0 blocks
==790619==    indirectly lost: 0 bytes in 0 blocks
==790619==      possibly lost: 0 bytes in 0 blocks
==790619==    still reachable: 334,490 bytes in 2,496 blocks
==790619==         suppressed: 1,827,238 bytes in 18,589 blocks
==790619== Reachable blocks (those to which a pointer was found) are not shown.
==790619== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==790619== 
==790619== For lists of detected and suppressed errors, rerun with: -s
==790619== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Sweet!

**Comment by jcupitt** [comment]
> Guard the whole dzsave operation with `#ifdef HAVE_ZIP`

I agree, this would be cleaner.

**Comment by kleisauke** [comment]
Great! I merged your locking branch (+ reverted commit 0ef43fd943df1fef95d16f8a128c0c0d2e448077 now that zipping is single-threaded).

> > Guard the whole dzsave operation with `#ifdef HAVE_ZIP`
> 
> I agree, this would be cleaner.

Done with commit 92d84163101c6c6401c130fa44f21edafdd13d04.

**Comment by kleisauke** [comment]
It seems that libgsf was always writing a ZIP64 file, even for small image pyramids:
```console
$ vips dzsave test/test-suite/images/sample.jpg x.zip
$ file x.zip 
x.zip: Zip archive data, at least v4.5 to extract, compression method=store
```

Whereas libzip does this automatically according to the docs of [`zip_file_extra_field_set()`](
https://libzip.org/documentation/zip_file_extra_field_set.html).

> Please note that the extra field IDs 0x0001 (ZIP64 extension), 0x6375 (Infozip UTF-8 comment), and 0x7075 (Infozip UTF-8 file name) can not be set using `zip_file_extra_field_set`() since they are set by [libzip(3)](https://libzip.org/documentation/libzip.html) automatically when needed.

```console
$ vips dzsave test/test-suite/images/sample.jpg x.zip
$ file x.zip 
x.zip: Zip archive data, at least v2.0 to extract, compression method=store
```

**Comment by kleisauke** [comment]
> - [ ] Test unseekable targets (e.g. pipes).

It seems that unseekable targets aren't working. :(

<details>
  <summary>Details</summary>

```python
import pyvips

output_file = open("x.zip", "w+b")


def write_handler(chunk):
    return output_file.write(chunk)


def read_handler(size):
    return output_file.read(size)


def seek_handler(offset, whence):
    output_file.seek(offset, whence)
    return output_file.tell()


def end_handler():
    try:
        output_file.close()
    except IOError:
        return -1
    else:
        return 0


target = pyvips.TargetCustom()
target.on_write(write_handler)
# target.on_read(read_handler)
# target.on_seek(seek_handler)
target.on_end(end_handler)

image = pyvips.Image.black(1000, 1000)
image.dzsave_target(target)
```
</details>

This means that the semantics of `dzsave_target` is the same as `tiffsave_target`, which also requires a target seek handler. Perhaps this is fine, although the previous libgsf-based saver did not require this.

**Comment by jcupitt** [comment]
> It seems that unseekable targets aren't working. 

That's strange. Could this be related to automatic zip32 / 64 support?

**Comment by kleisauke** [comment]
> Could this be related to automatic zip32 / 64 support?

It could be related, I remembered that we always wrote 'streamed' ZIP files when building against libgsf >= 1.14.31, see:
https://github.com/libvips/libvips/pull/2927#issuecomment-1184621317

Though, some unzip libraries fails on these type of ZIP files, see e.g.:
https://github.com/thejoshwolfe/yauzl#no-streaming-unzip-api

I'll investigate.

**Comment by kleisauke** [comment]
I just opened issue https://github.com/nih-at/libzip/issues/378 upstream.

**Comment by kleisauke** [comment]
I switched to the BSD 2-Clause licensed libarchive since libzip doesn't support unseekable targets. 

Here's a benchmark on my new(!) Ryzen 9 7900 workstation:

libgsf - af0c9c8e415cad32ed9946b23b46313b4f1841b5:
```console
$ /usr/bin/time -f %M:%e vips dzsave CMU-2.svs[attach-associated] x.szi --overlap 0 --suffix .png[compression=0] --vips-leak
vips_threadset_free: peak of 105 threads
memory: high-water mark 728.43 MB
1283940:23.88
```

libzip - e76251aedf13d6ea470ccef22a0223ce9563fd3c:
```console
$ /usr/bin/time -f %M:%e vips dzsave CMU-2.svs[attach-associated] x.szi --overlap 0 --suffix .png[compression=0] --vips-leak
vips_threadset_free: peak of 105 threads
memory: high-water mark 718.22 MB
13558392:25.96
```

libarchive - f8eaf631335f30151de91dbddd9e525667ba2934:
```console
$ /usr/bin/time -f %M:%e vips dzsave CMU-2.svs[attach-associated] x.szi --overlap 0 --suffix .png[compression=0] --vips-leak
vips_threadset_free: peak of 77 threads
memory: high-water mark 669.83 MB
1078368:19.67
```

This PR is ready for review now.

**Review by lovell — CHANGES_REQUESTED** [review]
Thanks Kleis, I've left one suggestion inline to improve compatibility of the zip files produced, but otherwise LGTM.

File: libvips/foreign/dzsave.c [hunk 2]
```diff
@@ -2532,87 +2207,63 @@ vips_foreign_save_dz_build( VipsObject *object )
...
+		char compression_string[2] = { '0' + dz->compression, 0 };
+		if( archive_write_set_format_option( dz->archive, "zip",
+			"compression-level", compression_string ) != ARCHIVE_OK ) {
+			archive_write_free( dz->archive );
+			return( -1 );
 		}
-#endif /*HAVE_GSF_DEFLATE_LEVEL*/
 
-		dz->tree = vips_gsf_tree_new( out2, dz->compression );
+		/*if( archive_write_add_filter( dz->archive,
```
[the remark this answers is not in the document]
Comment (kleisauke) [reply libvips/foreign/dzsave.c:2271]: Great! It looks like this also fixes the file size issue in libvips' test suite.

Commit f1120af390e782e8d60dc660c372aed1520a80a9 incorporates this.

File: libvips/foreign/dzsave.c [hunk 3]
```diff
@@ -2532,87 +2207,80 @@ vips_foreign_save_dz_build( VipsObject *object )
...
+			return( -1 );
+		}
 
-	default:
-		g_assert_not_reached();
+		/*if( archive_write_add_filter( dz->archive,
+			ARCHIVE_FILTER_NONE ) != ARCHIVE_OK ) {
+			archive_write_free( dz->archive );
+			return( -1 );
+		}*/
```
Comment (lovell) [inline libvips/foreign/dzsave.c:2275]: Can this commented-out block be removed?

Comment (kleisauke) [reply libvips/foreign/dzsave.c]: aee3605c9cace5f5941cbdbf247113e3098c89fc

**Review by lovell — APPROVED** [review]
:+1:

**Comment by kleisauke** [comment]
Commit https://github.com/libvips/build-win64-mxe/commit/eb98ddf510c2000b74ec448bfbd5859428e8fb70 backports this within the Windows builds for the upcoming v8.14.3 release.

**Comment by lovell** [comment]
Thanks Kleis, I used your patch to backport this for sharp's Linux and macOS binaries too - see https://github.com/lovell/sharp-libvips/commit/ac464fc975ad123f875c606ee5b8b2fee3144d91

**Pull request merged**